#!/bin/bash
set -e
#########################################################
#	SCRIPT TO FILTER BAC READS USING BOWTIE2 MAPPING	#
#########################################################
# Arguments:
# $1 = bacDB. File with the reference genome for the mapping. Must be adjacent to the bowtie index files.
# $2 = sampleName. Name of the sample to be processed. Must match the name of the sample in the RAW directory.
# 1. Creates necessary directories. 
# 2. Maps against bacteria reference genome.
# 3. Creates .fastq files using the host-free files generated by host_removal.sh
# Output files: (In ANALYSIS/sampleName/03.BACTERIA/)
# sampleName_bacteria_mapped.sam: SAM file from mapping the processed files against the reference genome.
# sampleName_Bacteria_R1.fastq: .fastq file with R1 reads that mapped the bacteria DB.
# sampleName_Bacteria_R2.fastq: .fastq file with R2 reads that mapped the bacetria DB.
# sampleName_bacteria_mapping.log: .log file with a log of the mapping.

function map_bacteria {
#	GET ARGUMENTS
bacDB=$1  
sampleAnalysisDir=$2
#	INITIALIZE VARIABLES
#		Directories
sampleName=$(basename "${sampleAnalysisDir}")
bacFilesDir="${sampleAnalysisDir}/03.BACTERIA/" #directory where the host filtering files will we saved (sam for mapping and fastq for host free samples)
noHostDir="${sampleAnalysisDir}/02.HOST/"
#		Input Files
noHostR1Fastq="${noHostDir}${sampleName}_noHost_R1.fastq"
noHostR2Fastq="${noHostDir}${sampleName}_noHost_R2.fastq"
#		OutputFiles
bowtie2logFile="${bacFilesDir}${sampleName}_bacteria_mapping.log"
BacMappedR1Fastq="${bacFilesDir}${sampleName}_Bacteria_R1.fastq"
BacMappedR2Fastq="${bacFilesDir}${sampleName}_Bacteria_R2.fastq"
mappedSamFile="${bacFilesDir}${sampleName}_bacteria_mapped.sam"

echo -e "$(date)" 
echo -e "*********** MAPPING BACTERIA IN $sampleName ************"

#	CREATE DIRECTORY FOR THE SAMPLE IF NECESSARY
if [ ! -d ${bacFilesDir} ]
then
	mkdir -p $bacFilesDir
	echo -e "${bacFilesDir} created"
fi
	
#	BOWTIE2 MAPPING AGAINST BACTERIA	
echo -e "--------Bowtie2 is mapping against the reference ....------"
echo -e "$(date)\t Start mapping ${sampleName}\n" > $bowtie2logFile
echo -e "The command is: ### bowtie2 -fr -x "$bacDB" -q -1 $noHostR1Fastq -2 $noHostR2Fastq -S $mappedSamFile ###\n" >> $bowtie2logFile 
bowtie2 -fr -x "$bacDB" -q -1 $noHostR1Fastq -2 $noHostR2Fastq -S $mappedSamFile 2>&1 | tee -a $bowtie2logFile
echo -e "$(date)\t Finished mapping ${sampleName}\n" >> $bowtie2logFile

#	SEPARATE R1 AND R2 MAPPED READS AND FILTER HOST
echo -e "----------------- Filtering bacteria reads ...---------------------"
echo -e "$(date)\t Start filtering ${sampleName}\n" >> $bowtie2logFile
#echo -e "The command is: ###samtools view -F 0x40 $mappedSamFile | awk '{if($3 == '*') print '@'$1'\\n'$10'\\n''+'$1'\\n'$11}' > $noHostR1Fastq" >> $bowtie2logFile
echo -e "The command is: ###samtools view -F 0x40 $mappedSamFile | awk '{if($3 != '*') print '@' $1 '\\n' $10 '\\n' '+' '\\n' $11}' > $BacMappedR1Fastq" >> $bowtie2logFile
samtools view -F 0x40 $mappedSamFile | awk '{if($3 != "*") print "@" $1 "\n" $10 "\n" "+" $1 "\n" $11}' > $BacMappedR1Fastq
echo -e "The command is: ###samtools view -f 0x40 $mappedSamFile | awk '{if($3 != '*') print '@' $1 '\\n' $10 '\\n' '-' '\\n' $11}' > $BacMappedR2Fastq" >> $bowtie2logFile
samtools view -f 0x40 $mappedSamFile | awk '{if($3 != "*") print "@" $1 "\n" $10 "\n" "+" $1 "\n" $11}' > $BacMappedR2Fastq
#	samtools separates R1 (-F) or R2 (-f) reads using the mapped SAM file and awk filters those mapped (=!"*") in fastq format
echo -e "$(date)\t Finished filtering ${sampleName}\n" >> $bowtie2logFile

echo -e "$(date)" 
echo -e "*********** FINISHED MAPPING BACTERIA IN $sampleName ************"
}
#map_bacteria /processing_Data/bioinformatics/research/20160530_METAGENOMICS_AR_IC_T/REFERENCES/BACTERIA_16S_REFERENCE/16S /processing_Data/bioinformatics/research/20160530_METAGENOMICS_AR_IC_T/ANALYSIS/MuestraPrueba/
