#########################################################
#		  		SCRIPT TO MERGE RESULTS				 	#
#########################################################

# 1. Reads coverage and blast files. 
# 2. Generates txt file with the merged table.
# Note: This script should only be run after finishing the blast and coverage analysis of the sample for that organism. 

# Arguments:
# $1 (sampleName) = Name of the sample. (sampleName)
# $2 (organism) = Name of the organism directory (xx-organism)

# Input Files:
# sampleName_coverageTable.txt: File generated with graphs_coverage.R (In ANALYSIS/xx-organism/sampleName/coverage/)
# sampleName_BLASTn_filtered.blast: File generated by the blast script. (In ANALYSIS/xx-organism/sampleName/blast/)

# Output files: (In RESULTS/data/)
# sampleName_organism_results.txt: txt file of the merged results table.




# ARGUMENTS
args = commandArgs(trailingOnly=TRUE)
sampleName=args[1] # sampleName
organism=args[2] # xx-organism

# CONSTANTS
workingDir="/processing_Data/bioinformatics/research/20160530_METAGENOMICS_AR_IC_T/"
sampleCoverageTable=paste(workingDir, "ANALYSIS/", organism, "/", sampleName, "/coverage/", sampleName,"_coverageTable.txt", sep='')
sampleBlastTable=paste(workingDir, "ANALYSIS/", organism, "/", sampleName, "/blast/",sampleName, "_BLASTn_filtered.blast", sep= '')

# READ INPUT FILES
coverage=read.table(sampleCoverageTable, sep="\t", header=TRUE)
blast=read.table(sampleBlastTable, sep="\t", header=FALSE)

# 	Name input files 
colnames(blast) <- c("Organism","Query_seq_id","Reference Id","% of identical matches","Alignment length", "Number of mismatches", 
					 "Number of Gap openings", "Start of alignment in query", "End of alignment in query", "Start of alignment in subject", 
					 "Expect value", "Bit Score", "Query")

# 	dropping query id and query sequence
blast = subset(blast, select = c("Organism","Reference Id","% of identical matches","Alignment length", "Number of mismatches",
								 "Number of Gap openings", "Start of alignment in query", "End of alignment in query", "Start of alignment in subject",
								 "Expect value", "Bit Score"))

# 	merge files
sampleResults = merge(x=blast, y=coverage, by.x = "Reference Id", by.y = "gnm")

# WRITE AOUTPUT FILE WITH MERGED TABLES
organism=unlist(strsplit(organism, split='-', fixed=TRUE))[2]
write.table(sampleResults, file=(paste(workingDir, "RESULTS/data/persamples/", sampleName, "_", organism, "_results.txt", sep="")), sep= '\t', col.names=FALSE, row.names=FALSE)


